### ISSUE TITLE
Feature request for [AlertDialog]

### ISSUE THREAD

**Issue by hyochan** [dooboolab/dooboo-ui#71]
**Is your feature request related to a problem? Please describe.**
It'd be good to have `dooboo-ui`'s own `AlertDialog` or we should use the non fancy default dialog.

**Describe the solution you'd like**
<img width="148" alt="Screen Shot 2021-08-16 at 7 16 55 PM" src="https://user-images.githubusercontent.com/27461460/129548722-3d223577-2265-4fcc-a2b9-0d642fa07499.png">
<img width="186" alt="Screen Shot 2021-08-16 at 7 16 37 PM" src="https://user-images.githubusercontent.com/27461460/129548758-824421bd-b8ec-430d-a940-c1a0776576a5.png">

**Additional context**
Above can be implemented in `@dooboo-ui/alert-dialog` as a seperate package because I think we might need [react-native-modalbox](https://github.com/maxs15/react-native-modalbox).

**Comment by do02reen24** [comment]
I want to develop this task.
Can you assign it to me?

**Comment by yujong-lee** [comment]
I look forward to your work!

As you know much better than I do, I leave a comment hoping that my little experience will help.

- - -

As @hyochan mentioned, it can be implemented in `@dooboo-ui/alert-dialog`.  Actually,  it is encouraging if you use `react-native-modalbox`. You can read [this](https://github.com/dooboolab/dooboo-ui/pull/49#issuecomment-894665037) for detail if you haven't.

So you can start developing `lerna create alert-dialog` and `lerna add react-native-modalbox --scope=alert-dialog`.

- - -

In `react-native-modalbox`, there are two ways to manage open/close state of modal.
1. `isOpen` prop. (recommended)
2. `open/close` method

You won't need second form, but just in case you use it, you need to do like below. (there's no documentation as I know.)

```ts
export default function AlertDialog(props: Props): ReactElement {
  const modal = useRef(null);

  return (
    <>
      <Button title="Closer" onPress={() => modal.current.close()} />,
      <Modal ref={modal} {...props} />
    </>
  );
}
```
- - -

Please let me know if I know something wrong.  
Hope this helps.

**Comment by yujong-lee** [comment]
@do02reen24 
Looking forward to your work!

I'm sure you know better than me, but I'll share some source just in case it helps.

1. Reading [this](https://github.com/dooboolab/dooboo-ui/pull/49#issuecomment-894665037) will be helpful.
2. If you're going to implement in `@dooboo-ui/alert-dialog`,  you can start with `lerna create alert-dialog` and `lerna add react-native-modalbox --scope=alert-dialog`.
3. There are two ways to manage open state of modal in `react-native-modalbox`. 
 - `opened` props. (recommended)
 - `open/close` method
 You won't need second way, but just in case you want it, here's some code example. (there's no example of this in docs)

```ts
export default function AlertDialog(props: Props): ReactElement {
  const modal = useRef(null);

  return (
    <>
      <Button title="Closer" onPress={() => modal.current.close()} />,
      <Modal ref={modal} {...props} />
    </>
  );
}
```

Please let me know if I know something wrong.

Hope this helps!

**Comment by hyochan** [comment]
> I want to develop this task.
> Can you assign it to me?

Thanks for your interest~! I've just assigned you here.
Actually, you can start working on this even if you are not assigned. In the future, you can just tell us that you are working on it if you want to share your progress.

@yujong-lee Thanks for sharing the spec. This looks really useful!

**Comment by doreen0224** [comment]
I've summarized the props that AlertDialog needs.
Is there any option to add, modify, or remove? Please give me feedback.

  | types | default
-- | -- | --
isOpen? | boolean | false
type? | "alert" \|"confirm" \|"prompt" | "alert"
color? | string | "#000000"
modalStyle? | StyleProp<ViewStyle> | undefined
backdropOpacity? | Number | 0.5
backdropPressToClose? | boolean | true
renderContent? | (string) => JSX.Element | (string) => JSX.Element
onPress | (result: any) => void | -

**Comment by yujong-lee** [comment]
@doreen0224 All component in `dooboo-ui` should follow this rule for styling.

> Never make style props as another props.

@hyochan [mentioned it](https://github.com/dooboolab/dooboo-ui/pull/73#discussion_r692972595).


**So we need `style` for referring itself, and `styles` for wrapping things like `modalStyle`.**

#69 is my work related to this.

See [this comment](https://github.com/dooboolab/dooboo-ui/pull/69#discussion_r690120272) by @hyochan  too.

Code below is `Button`'s prop written by @hyochan .

```ts
type Styles = {
  container?: StyleProp<ViewStyle>;
  text?: StyleProp<TextStyle>;
  disabledButton?: StyleProp<ViewStyle>;
  disabledText?: StyleProp<TextStyle>;
  hovered?: StyleProp<ViewStyle>;
};

export interface ButtonProps {
  testID?: string;
  indicatorColor?: string;
  loading?: boolean;
  disabled?: boolean;
  outlined?: boolean;
  style?: StyleProp<ViewStyle>;
  styles?: Styles;
  leftElement?: ReactElement;
  rightElement?: ReactElement;
  activeOpacity?: TouchableOpacityProps['activeOpacity'];
  text?: string;
  onPress?: TouchableOpacityProps['onPress'];
  touchableOpacityProps?: Partial<TouchableOpacityProps>;
  textProps?: Partial<TextProps>;
  type?: ButtonType;
  size?: ButtonSize;
}
```

**Comment by yujong-lee** [comment]
@doreen0224  I think you missed `data: string[]` for `renderContent`?

**Comment by hyochan** [comment]
The [AlertDialog] might be used globally so you might want to consider support for context provider which we had in [Snackbar Legacy](https://github.com/dooboolab/dooboo-ui-legacy/blob/master/packages/Snackbar/index.tsx).

In out current [SnackBar](https://github.com/dooboolab/dooboo-ui/blob/master/main/Snackbar/index.tsx), we don't have the context provider and I just thought we need this support for the snackbar.

Thinking further, I think the [AlertDialog] is enough to be provided as `ref` unlike the [Snackbar] because it is not necessary to run on multiple screens simultaneouly. The snackbar should be opened even when page is expired (if this is hard to understand, we can have a sync!).

As said in the `description`, you can use `react-native-modalbox` but then it should be written in separate package in [package directory](https://github.com/dooboolab/dooboo-ui/tree/master/packages) since we need extra dependency. This is one of our policy (still not explicitly written in our document :sweat:) on managing packages. I've also [wrote a reply](https://github.com/dooboolab/dooboo-ui/pull/49#issuecomment-894665037) on the current context. Hope this helps!

Please tell if there are any further questions.

**Comment by hyochan** [comment]
Closes via #87. Wasn't linked properly.
